exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -218,4 +302,141 @@ private void registerOperatorsWithoutInference(DrillOperatorTable operatorTable)
+        }
+      }
+    }
+
+    public void removeJar(T jName) {
+      try (AutoCloseableLock lock = writeLock.open()) {
+        removeAllByJar(jName);
+      }
+    }
+
+    public List<T> getAllJarNames() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        return Lists.newArrayList(jars.keySet());
+      }
+    }
+
+    public List<T> getAllFunctionNames(T jName) {
+      try  (AutoCloseableLock lock = readLock.open()){
+        Map<T, List<T>> map = jars.get(jName);
+        return map == null ? Lists.<T>newArrayList() : Lists.newArrayList(map.keySet());
+      }
+    }
+
+    public ListMultimap<T, U> getAllFunctionsWithHolders() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        ListMultimap<T, U> multimap = ArrayListMultimap.create();
+        for (Entry<T, Map<T, U>> entry : functions.entrySet()) {
+          multimap.putAll(entry.getKey(), Lists.newArrayList(entry.getValue().values()));
+        }
+        return multimap;
+      }
+    }
+
+    public ListMultimap<T, T> getAllFunctionsWithSignatures() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        ListMultimap<T, T> multimap = ArrayListMultimap.create();
+        for (Entry<T, Map<T, U>> entry : functions.entrySet()) {
+          multimap.putAll(entry.getKey(), Lists.newArrayList(entry.getValue().keySet()));
+        }
+        return multimap;
+      }
+    }
+
+    public List<U> getHoldersByFunctionName(T fName) {
+      try (AutoCloseableLock lock = readLock.open()) {
+        Map<T, U> map = functions.get(fName);
+        return map == null ? Lists.<U>newArrayList() : Lists.newArrayList(map.values());
+      }
+    }
+
+    public boolean containsJar(T jName) {
+      try (AutoCloseableLock lock = readLock.open()) {
+        return jars.containsKey(jName);
+      }
+    }
+
+    public int functionsSize() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        return functions.size();
+      }
+    }
+
+    private void addFunction(T jName, T fName, T sName, U fHolder) {
+      Map<T, List<T>> map = jars.get(jName);
+
+      List<T> list = map.get(fName);
+      if (list == null) {
+        list = Lists.newArrayList();
+        map.put(fName, list);
+      }
+
+      if (!list.contains(sName)) {
+        list.add(sName);
+
+        Map<T, U> sigsMap = functions.get(fName);
+        if (sigsMap == null) {
+          sigsMap = Maps.newHashMap();
+          functions.put(fName, sigsMap);
+        }
+
+        U u = sigsMap.get(sName);
+        if (u == null) {
+          sigsMap.put(sName, fHolder);
+        }
+      }
+    }
+
+    private void removeAllByJar(T jName) {
+      Map<T, List<T>> removeMap = jars.remove(jName);
+      if (removeMap == null) {
+        return;
+      }
+
+      for (Map.Entry<T, List<T>> remEntry : removeMap.entrySet()) {
+        Map<T, U> fNameMap = functions.get(remEntry.getKey());
+        List<T> value = remEntry.getValue();
+        fNameMap.keySet().removeAll(value);
```

Comment:
Yes, removing key from keySet will the whole entry (key + value) from map, since keySet is backed by the map, so changes to the map are reflected in the set, and vice-versa.